plutus-chain-index/src/Plutus/ChainIndex/Lib.hs

```diff
@@ -131,14 +137,21 @@ toCardanoChainSyncHandler runReq handler = \case
 
 -- | A handler for chain synchronisation events.
 type ChainSyncHandler = ChainSyncEvent -> IO ()
+type BlocksChan = TChan CI.ChainSyncBlock
 
 -- | The default chain synchronisation event handler. Updates the in-memory state and the database.
 defaultChainSyncHandler :: RunRequirements -> ChainSyncHandler
 defaultChainSyncHandler runReq evt = void $ runChainIndexDuringSync runReq $ case evt of
-    (RollForward block _)  -> appendBlock block
+    (RollForward block _)  -> appendBlocks [block]
     (RollBackward point _) -> rollback point
     (Resume point)         -> resumeSync point
 
+storeChainSyncHandler :: BlocksChan -> RunRequirements -> ChainSyncHandler
+storeChainSyncHandler blocksChan runReq evt = case evt of
+    (RollForward block _)  -> atomically $ writeTChan blocksChan block
+    (RollBackward point _) -> void $ runChainIndexDuringSync runReq $ rollback point
+    (Resume point)         -> void $ runChainIndexDuringSync runReq $ resumeSync point
```

Comment:
Doing it this way, I think you have a concurrency problem.

My understanding of your code is that it writes blocks to the TChan and the appendBlocks is handled by a separate thread.

However, what might happen here is that the `RollBackward` event is called before the thread processing the `RollForward` blocks is actually finished. So, the `RollBackward` is called, it updates the db to a different point, and the thread processes blocks that should have been rollbacked.

I think the solution is to use the `ChainSyncEvent` type in the TChan, batch the `RollForward` events until a `RollBackward` is met and use `appendBlocks`. Something like that...

Reply:
Hmmm, yeah, I agree, IO operations with one resource should happen in one thread. I haven't thought much writing this, was trying to make the `appendBlocks` work. Will fix it. Thanks!

Reply:
>I think the solution is to use the ChainSyncEvent type in the TChan, batch the RollForward events until a RollBackward is met and use appendBlocks. Something like that...

Just reading `RollForward` events until `RollBackward` looks fragile as we keep a lot of `RollForward`s in memory and it's very easy to loose the processed state.